File: ontopy/ontologenius/ConvertionClient.py
import rospy

from ontologenius.srv import OntologeniusConvertion
from ontologenius.srv import OntologeniusConvertionRequest

class ConvertionClient:
    

    def __init__(self, name):
        """Constructs a ROS client linked to the service name(str)."""
        self._name = 'ontologenius/convertion'
        if name != '':
          self._name = self._name + name
        self._verbose = False
        self._client = rospy.ServiceProxy(self._name, OntologeniusConvertion, True)

    def setVerbose(self, verbose):
        """If verbose(bool) is set to True, the clients will post messages about
           the failure to call the services and about their restoration.
        """
        self._verbose = verbose

    def individualsIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConvertionRequest.INDIVIDUALS)

    def classesIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConvertionRequest.CLASSES)

    def dataPropertiesIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConvertionRequest.DATA_PROPERTIES)

    def objectPropertiesIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConvertionRequest.OBJECT_PROPERTIES)

    def literalsIndexes2Ids(self, indexes):
       return self._indexes2Ids(indexes, OntologeniusConvertionRequest.LITERAL)
    

    def individualsIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConvertionRequest.INDIVIDUALS)

    def classesIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConvertionRequest.CLASSES)

    def dataPropertiesIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConvertionRequest.DATA_PROPERTIES)

    def objectPropertiesIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConvertionRequest.OBJECT_PROPERTIES)

    def literalsIndex2Id(self, index):
       return self._index2Id(index, OntologeniusConvertionRequest.LITERAL)
    

    def individualsIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConvertionRequest.INDIVIDUALS)

    def classesIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConvertionRequest.CLASSES)

    def dataPropertiesIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConvertionRequest.DATA_PROPERTIES)

    def objectPropertiesIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConvertionRequest.OBJECT_PROPERTIES)

    def literalsIds2Indexes(self, ids):
       return self._ids2Indexes(ids, OntologeniusConvertionRequest.LITERAL)
    

    def individualsId2Index(self, id):
       return self._id2Index(id, OntologeniusConvertionRequest.INDIVIDUALS)

    def classesId2Index(self, id):
       return self._id2Index(id, OntologeniusConvertionRequest.CLASSES)

    def dataPropertiesId2Index(self, id):
       return self._id2Index(id, OntologeniusConvertionRequest.DATA_PROPERTIES)

    def objectPropertiesId2Index(self, id):
       return self._id2Index(id, OntologeniusConvertionRequest.OBJECT_PROPERTIES)

    def literalsId2Index(self, id):
       return self._id2Index(id, OntologeniusConvertionRequest.LITERAL)

    def _indexes2Ids(self, indexes, source):
        response = self._call(source, [], indexes)
        if response:
          return response.values_str
        else:
          return None

    def _index2Id(self, index, source):
        response = self._call(source, [], [index])
        if response:
          return response.values_str[0]
        else:
          return None

    def _ids2Indexes(self, ids, source):
        response = self._call(source, ids, [])
        if response:
          return response.values_int
        else:
          return None

    def _id2Index(self, id, source):
        response = self._call(source, [id], [])
        if response:
          return response.values_int[0]
        else:
          return None

    def _call(self, source, values_str, values_int):
        try:
            response = self._client(source, values_str, values_int)
            return response
        except (rospy.ServiceException, rospy.exceptions.TransportTerminated) as e:
            if self._verbose == True:
                print("Failure to call " + self._name)
            self._client = rospy.ServiceProxy(self._name, OntologeniusConvertion, True)
            try:
                response = self._client(source, values_str, values_int)
                if self._verbose == True:
                    print("Restored " + self._name)
                return response
            except (rospy.ServiceException, rospy.exceptions.TransportTerminated) as e:
                if self._verbose == True:
                    print("Failure of service restoration")
                return None

File: src/API/ontologenius/ConvertionClient.cpp
#include "ontologenius/API/ontologenius/ConvertionClient.h"

#ifndef COLOR_OFF
#define COLOR_OFF     "\x1B[0m"
#endif
#ifndef COLOR_RED
#define COLOR_RED     "\x1B[0;91m"
#endif
#ifndef COLOR_ORANGE
#define COLOR_ORANGE  "\x1B[1;33m"
#endif
#ifndef COLOR_GREEN
#define COLOR_GREEN   "\x1B[1;92m"
#endif

namespace onto {

ConvertionClient::ConvertionClient(ros::NodeHandle* n, const std::string& name) : name_((name == "") ? "/ontologenius/convertion" : "/ontologenius/convertion/" + name),
                                                                                  n_(n), verbose_(false),
                                                                                  client_(n->serviceClient<ontologenius::OntologeniusConvertion>(name_, true))
{}

std::vector<std::string> ConvertionClient::individualsIndex2Id(const std::vector<int64_t>& indexes)
{
  return index2Id(indexes, ontologenius::OntologeniusConvertionRequest::INDIVIDUALS);
}

std::vector<std::string> ConvertionClient::classesIndex2Id(const std::vector<int64_t>& indexes)
{
  return index2Id(indexes, ontologenius::OntologeniusConvertionRequest::CLASSES);
}

std::vector<std::string> ConvertionClient::dataPropertiesIndex2Id(const std::vector<int64_t>& indexes)
{
  return index2Id(indexes, ontologenius::OntologeniusConvertionRequest::DATA_PROPERTIES);
}

std::vector<std::string> ConvertionClient::objectPropertiesIndex2Id(const std::vector<int64_t>& indexes)
{
  return index2Id(indexes, ontologenius::OntologeniusConvertionRequest::OBJECT_PROPERTIES);
}

std::vector<std::string> ConvertionClient::literalsIndex2Id(const std::vector<int64_t>& indexes)
{
  return index2Id(indexes, ontologenius::OntologeniusConvertionRequest::LITERAL);
}

std::string ConvertionClient::individualsIndex2Id(int64_t index)
{
  return index2Id(index, ontologenius::OntologeniusConvertionRequest::INDIVIDUALS);
}

std::string ConvertionClient::classesIndex2Id(int64_t index)
{
  return index2Id(index, ontologenius::OntologeniusConvertionRequest::CLASSES);
}

std::string ConvertionClient::dataPropertiesIndex2Id(int64_t index)
{
  return index2Id(index, ontologenius::OntologeniusConvertionRequest::DATA_PROPERTIES);
}

std::string ConvertionClient::objectPropertiesIndex2Id(int64_t index)
{
  return index2Id(index, ontologenius::OntologeniusConvertionRequest::OBJECT_PROPERTIES);
}

std::string ConvertionClient::literalsIndex2Id(int64_t index)
{
  return index2Id(index, ontologenius::OntologeniusConvertionRequest::LITERAL);
}

std::vector<int64_t> ConvertionClient::individualsId2Index(const std::vector<std::string>& ids)
{
  return Id2Index(ids, ontologenius::OntologeniusConvertionRequest::INDIVIDUALS);
}

std::vector<int64_t> ConvertionClient::classesId2Index(const std::vector<std::string>& ids)
{
  return Id2Index(ids, ontologenius::OntologeniusConvertionRequest::CLASSES);
}

std::vector<int64_t> ConvertionClient::dataPropertiesId2Index(const std::vector<std::string>& ids)
{
  return Id2Index(ids, ontologenius::OntologeniusConvertionRequest::DATA_PROPERTIES);
}

std::vector<int64_t> ConvertionClient::objectPropertiesId2Index(const std::vector<std::string>& ids)
{
  return Id2Index(ids, ontologenius::OntologeniusConvertionRequest::OBJECT_PROPERTIES);
}

std::vector<int64_t> ConvertionClient::literalsId2Index(const std::vector<std::string>& ids)
{
  return Id2Index(ids, ontologenius::OntologeniusConvertionRequest::LITERAL);
}

int64_t ConvertionClient::individualsId2Index(const std::string& id)
{
  return Id2Index(id, ontologenius::OntologeniusConvertionRequest::INDIVIDUALS);
}

int64_t ConvertionClient::classesId2Index(const std::string& id)
{
  return Id2Index(id, ontologenius::OntologeniusConvertionRequest::CLASSES);
}

int64_t ConvertionClient::dataPropertiesId2Index(const std::string& id)
{
  return Id2Index(id, ontologenius::OntologeniusConvertionRequest::DATA_PROPERTIES);
}

int64_t ConvertionClient::objectPropertiesId2Index(const std::string& id)
{
  return Id2Index(id, ontologenius::OntologeniusConvertionRequest::OBJECT_PROPERTIES);
}

int64_t ConvertionClient::literalsId2Index(const std::string& id)
{
  return Id2Index(id, ontologenius::OntologeniusConvertionRequest::LITERAL);
}

std::vector<std::string> ConvertionClient::index2Id(const std::vector<int64_t>& indexes, int8_t source)
{
  ontologenius::OntologeniusConvertion srv;
  srv.request.source = source;
  srv.request.values_int = indexes;
  if(call(srv))
    return srv.response.values_str;
  else
    return {};
}

std::string ConvertionClient::index2Id(int64_t index, int8_t source)
{
  ontologenius::OntologeniusConvertion srv;
  srv.request.source = source;
  srv.request.values_int = {index};
  if(call(srv))
    return srv.response.values_str.front();
  else
    return "";
}

std::vector<int64_t> ConvertionClient::Id2Index(const std::vector<std::string>& ids, int8_t source)
{
  ontologenius::OntologeniusConvertion srv;
  srv.request.source = source;
  srv.request.values_str = ids;
  if(call(srv))
    return srv.response.values_int;
  else
    return {};
}

int64_t ConvertionClient::Id2Index(const std::string& id, int8_t source)
{
  ontologenius::OntologeniusConvertion srv;
  srv.request.source = source;
  srv.request.values_str = {id};
  if(call(srv))
    return srv.response.values_int.front();
  else
    return {};
}

bool ConvertionClient::call(ontologenius::OntologeniusConvertion& srv)
{
  if(client_.call(srv))
    return true;
  else
  {
    if(verbose_)
      std::cout << COLOR_ORANGE << "Failure to call " << name_ << COLOR_OFF << std::endl;
    client_ = n_->serviceClient<ontologenius::OntologeniusConvertion>(name_, true);
    if(client_.call(srv))
    {
      if(verbose_)
        std::cout << COLOR_GREEN << "Restored " << name_ << COLOR_OFF << std::endl;
      return true;
    }
    else
    {
      if(verbose_)
        std::cout << COLOR_RED << "Failure of service restoration" << COLOR_OFF << std::endl;
      return false;
    }
  }
}

} // namespace onto
